src/Catalog/Helpers/ParallelAsync.cs

```diff
@@ -3,25 +3,24 @@
 
 using System;
 using System.Linq;
+using System.Net;
 using System.Threading.Tasks;
 
 namespace NuGet.Services.Metadata.Catalog.Helpers
 {
     public static class ParallelAsync
     {
-        public const int DefaultDegreeOfParallelism = 32;
-
         /// <summary>
         /// Creates a number of tasks specified by <paramref name="degreeOfParallelism"/> using <paramref name="taskFactory"/> and then runs them in parallel.
         /// </summary>
         /// <param name="taskFactory">Creates each task to run.</param>
-        /// <param name="degreeOfParallelism">The number of tasks to create.</param>
+        /// <param name="degreeOfParallelism">The number of tasks to create. Defaults to <see cref="ServicePointManager.DefaultConnectionLimit"/></param>
         /// <returns>A task that completes when all tasks have completed.</returns>
-        public static Task Repeat(Func<Task> taskFactory, int degreeOfParallelism = DefaultDegreeOfParallelism)
+        public static Task Repeat(Func<Task> taskFactory, int? degreeOfParallelism = null)
         {
             return Task.WhenAll(
                 Enumerable
-                    .Repeat(taskFactory, degreeOfParallelism)
+                    .Repeat(taskFactory, degreeOfParallelism ?? ServicePointManager.DefaultConnectionLimit)
```

Comment:
Should we be using `ServicePointManager.DefaultConnectionLimit`? This setting is a limit on TCP connections, but it looks like you're using it for limiting the number of storage queue messages which is completely different.

Based on your description, is the real issue that the same failed packages keep getting requeued? Should we instead introduce a cache to avoid requeing the same packages?

Reply:
This was an additional parallelism improvement I added, unrelated to the primary purpose of this PR. My thoughts are that the number of messages being processed simultaneously should be allowed to be as great as the number of web connections we can make at the same time. I can revert this change if you want, I didn't think it was very risky/notable.